### ISSUE TITLE
Coupling matrices used in iCubGenova04 configuration files

### ISSUE THREAD

**Issue by traversaro** [robotology/robots-configuration#39]
After the recent improvements in the configuration files structure for iCub robots documented in https://github.com/robotology/icub-main/blob/master/src/doc/release/v1_8_0.md#configuration-files , I briefly tried to validate the coupling matrices for the iCub shoulders and torso committed in this repo for iCubGenova04 and contained in these files: 
* Left Shoulder: https://github.com/robotology/robots-configuration/blob/0ff6cee7bd4dee7811811e4d2eee29b16407234b/iCubGenova04/hardware/mechanicals/left_arm-eb1-j0_3-mec.xml#L43
* Right Shoulder: https://github.com/robotology/robots-configuration/blob/0ff6cee7bd4dee7811811e4d2eee29b16407234b/iCubGenova04/hardware/mechanicals/right_arm-eb3-j0_3-mec.xml#L44
* Torso : https://github.com/robotology/robots-configuration/blob/0ff6cee7bd4dee7811811e4d2eee29b16407234b/iCubGenova04/hardware/mechanicals/torso-eb5-j0_2-mec.xml#L40  
and the coupling matrices documented in http://wiki.icub.org/wiki/ICub_coupled_joints , but I was unable to to do so. 

In particular I have several findings/doubts. 

# Matrix serialization if the AxisMap argument is used 
It is not clear the serialization used in the coupling matrix for the boards (like the torso) that have their axis remapped using the `AxisMap` argument. 

# Semantics of the matrixJ2M and matrixM2J parameters 
It is not clear the meaning of the `matrixJ2M` (and `matrixM2J`) parameter, i.e. if it is the matrix mapping joint velocities to motor velocities (motor velocities to joint velocities) or the matrix mapping joint torques to motor torques (motor torques to joint torques). 
The point of the code in which this variables are not clear to me enough to infer their meaning: 
* https://github.com/robotology/icub-main/blob/2fe40bcad13734ce6800839a0492c5bae99f3cb6/src/libraries/icubmod/embObjMotionControl/embObjMotionControl.cpp#L1033
* https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/Controller.c#L463
* https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L141
* https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L575
* https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L597

# Coupling matrix actually used in the configuration files 
Trying to match the matrix documented in http://wiki.icub.org/wiki/ICub_coupled_joints with the one available in the configuration files, it seems that at least for the shoulder, the `matrixM2J` is the matrix mapping the motor velocities to the joint velocities. However, it is not clear what is the `matrixJ2M`, that is not the inverse, neither the transpose nor the inverse transpose of the `matrixM2J` parameter. 

How have been this matrices chosen? 

Tagging a few people that could provide precious information: 
@valegagge @ale-git @marcoaccame @randaz81 @lornat75 @nunoguedelha @S-Dafarra @DanielePucci @francesco-romano @fiorisi .

**Comment by ale-git** [comment]
I suggest a short meeting to clarify these topics :)

**Comment by valegagge** [comment]
I would like give you some basic information.

First of all, the metrices written in configuration files are the same matrices that have been used in firmware from a lot of time.
The changes done in configuration files of version 1.8.0  (https://github.com/robotology/icub-main/blob/master/src/doc/release/v1_8_0.md#configuration-files ) have the objective to make firmware as configuralble as possible and to simplify firmware mantainance.

Regarding the `AxisMap` parameter, it transforms joint  logic number to hardware number. The firmware works with joint hardware number, therefore the matrices are written using joint hardware number.

The semantic matrix parameters are the following:
-  `MatrixJ2M` : transforms a measure from joints space to motor space
- `MatrixM2J` :   transforms a measure from joints space to motor space.
In the meeting we can study in deep their meaning and how the firmware code use it.

I know this post is not exhaustive, but I hope that in the meeting we can clarify any doubt.

**Comment by ale-git** [comment]
Waiting for the meeting, some explainations about the matrix semantic:

**matrixM2J** is the Jacobian matrix that transforms the velocities expressed in the motor space into the velocities expressed in the joint space, i.e. matrixM2J = dqj/dqm. It depends exclusively on the mechanical implementation of the joint set, and can't be modified. It is used in the code in two points:

1) to transform the motor encoder velocity measures into joint velocities https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L141

2) its transposed matrix matrixM2J' is used to convert torque measures and torque references from joint space to motor space, in fact torque control is performed at motor level https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L575  

**matrixJ2M** isn't, generally speaking, the inverse of matrixM2J. It is used in position/velocity stiff control, performed at joint level and decoupled after the control loops https://github.com/robotology/icub-firmware/blob/f5850b0d935e814eab5f1a448a6471b141dfcc5f/emBODY/eBcode/arch-arm/embobj/plus/mc/JointSet.c#L597

Since the output of the decoupling section isn't torque reference nor velocity reference, but PWM reference for the motors, what matrixJ2M should be it's still an open issue. If we had torque controlled motors at very low level, then matrixJ2M should be matrixM2J transposed. If we had velocity controlled motors at very low level, matrixJ2M should be the inverse of matrixM2J. With R1 robot we are exactly in this configuration, but both matrixM2J and matrixJ2M are identity matrix because we have no velocity controlled coupled joints. (Note: at present, we have no low level velocity controlled robot with coupled joints, but I tested an experimental configuration on a iCub robot, with 2FOC boards configured to perform inner velocity control, and it works very well). But, as mentioned above, with PWM reference output it isn't clear what we should use... theoretically, any matrixJ2M that satisfies the condition "matrixM2J*matrixJ2M is positive defined" shoud be able to control a joint set in stiff mode.

The matrixJ2M for waist and shoulder of iCub are configured as if the third joint were independent, so that the position PID controller of the third joint deals with the coupling like a disturbance. This is the reason why matrixJ2M isn't the inverse of matrixM2J for waist and shoulder. Any suggestion about "with PWM reference controlled motor matrixJ2M should be ..." is very welcome! :)

**Comment by traversaro** [comment]
Thanks for the comment @ale-git ! I think that for the moment this is everything we need to know from the point of view of friction estimation. I think that it would be convenient to have this information in the documentation of the `embObjMotionControl`, but that would be an issue in the `icub-main` repo, so I think we can close the issue here.

**Comment by nunoguedelha** [comment]
Thanks for the information @ale-git and @valegagge . Actually, for estimating the joint friction and motor parameters (`Ktau`, `Kv`, `Kc`), we need the original jacobian matrix **matrixM2J** transforming the velocities from the motor space to the joint space.
The variable `kinematic_mj` which value we can currently retrieve through the interface `IRemoteVariables` holds the values of **matrixJ2M**.

We would need 2 changes:
1. having an additional variable, accessible through the same interface, for instance named `kinematic_jm` or more explicitely `couplingMxJ2M` holding the actual jacobian matrix **matrixM2J**
2. a list of motor names (as we have one for joint names), instead of just indexes, and makes it easier to match the software with the actual physical configuration
3. two methods `getMotorNames()` and `getJointNames()` (`IAxisInfo` interface) for returning the respective lists
4. the jacobian matrix **matrixM2J** should be formatted following the order of the motors and joints measurements in the YARP port vector `/icub/<part>/stateExt:o`, that order being the same as the one in the lists returned by `getMotorNames()` and `getJointNames()`
I also suggest to change the name of `matrixJ2M` to something else in order to avoid confusion, or someone sooner or later coming back to you asking why this matrix is not the invert of etc..
for instance **couplingMxJ2MforCtrl**.

**Comment by nunoguedelha** [comment]
Just a heads up...
trying again to match the matrix documented in http://wiki.icub.org/wiki/ICub_coupled_joints with the one defined in the configuration files, there is a mismatch for the left **shoulder** case..
The documented matrix is:
![image](https://user-images.githubusercontent.com/6848872/29996741-8dafeabe-9004-11e7-8e13-f78907ebaf29.png)
(with t=0.615)

The one defined in the configuration file is:
https://github.com/robotology/robots-configuration/blob/0ff6cee7bd4dee7811811e4d2eee29b16407234b/iCubGenova04/hardware/mechanicals/left_arm-eb1-j0_3-mec.xml#L51-L56

**=> All the axes orientations are inverted**.

I suggest we keep this issue open until everything is clarified. Eventually some parameters will be updated to make the interface more straightforward and free of ambiguities.

**Comment by valegagge** [comment]
Hi @nunoguedelha ,

I saw your posts and tray to answer for points:

1) ok, but I have a dubt: is there a yarp interface that best suits matrices info? I know remoteInterface should be a way to get raw value, (value of firmware) and not value at user level.

2) The name of motor change according to version of eletronics and robot wiring. Could you explain me why do you need their names?

4) I think it is not a good idea, because each remote variable returns joint not ordered with AxisMap, in other words, returns value like firmware order.

Regarding of renamingof  `matrixJ2M` in couplingMxJ2MforCtrl it could be a good idea. Maybe we need also what is its real meaning... So if it is a parameter depending on control type, maybe we could move it from mecchanical files to motor control files.

At the end, regarding of mismatch sign of matrix of left shoulder, i think it is due to the new parameter `Gearbox_M2J`: its absolute value is the total gearbox between motor and joint, while the sign is used to mantain axes orientation.
In other words, mechanics of the two arms are equal, they differs in rotation of motor, so we need to invert sign to mantain axises orientation.
Maybe @ale-git could you provide more in-depth information.

**Comment by nunoguedelha** [comment]
Thanks @valegagge 
### regarding the requirements for **matrixM2J** retrieval
1. you're absolutely right, the current interface isn't appropriate, we were using it just as a quick dirty solution. A more suitable interface would be `IMotor` or `IAxisInfo`, and would implemented by `RemoteControlBoard` and `RemoteControlBoardRemapper`.
2. We don't need the real names of the motors, but some label associating a motor to its relative position in the coupling. The reasons are:
- in the upgraded friction calibration procedure we just started testing, we need to control the motors individually even in the case of coupled joints. In case some motor has a bad behaviour and requires particular attention, it's useful to be able to identify it on the hardware without ambiguity.
- as the calibration for each motor is saved over time, it is safer to associate that calibration with a label, tied to the relative position of the motor in the coupling. In case a motor is replaced, and a new calibration required, it is easier to control the results of that calibration.
- in case of doubts as the ones discussed in this thread, it is straightforward to look at the documentation and compare the labels there with the ones we get from the interface.
3. let's focus on `IAxisInfo` user interface.. the idea is to get from that interface a list of joints and a list of motor labels (independent from any `AxisMap` mapping handled by the firmware), along with the coupling matrix. The elements `X_jm` order in the coupling matrix would be mapped by the motors / joints lists returned by `IAxisInfo` interface.

### regarding of mismatch sign of matrix of left shoulder
ok, in this case, the total transform `Gearbox_M2J` x `matrixM2J` is inverted for both shoulders. We should probably just update the documentation.

**Comment by traversaro** [comment]
cc @gabrielenava

**Comment by pattacini** [comment]
iCubGenova04 is in icebox ❄️
